src/core/net/dns_client.cpp

```diff
@@ -350,13 +350,29 @@ Error Client::AddressResponse::GetAddress(uint16_t aIndex, Ip6::Address &aAddres
 
 Error Client::AddressResponse::GetNat64Prefix(Ip6::Prefix &aPrefix) const
 {
-    // Use well-known prefix "64:ff9b::/96" (temporary solution).
-    static const uint8_t kWellknownPrefix[] = {0x00, 0x64, 0xff, 0x9b, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+    Error                            error      = kErrorNotFound;
+    NetworkData::Iterator            iterator   = NetworkData::kIteratorInit;
+    signed int                       preference = OT_ROUTE_PREFERENCE_LOW;
+    NetworkData::ExternalRouteConfig config;
 
     aPrefix.Clear();
-    aPrefix.Set(kWellknownPrefix, 96);
 
-    return kErrorNone;
+    while (mInstance->Get<NetworkData::Leader>().GetNextExternalRoute(iterator, config) == kErrorNone)
+    {
+        if (!config.mNat64)
+        {
+            continue;
+        }
+
+        if ((aPrefix.GetLength() == 0) || (config.mPreference > preference))
+        {
+            aPrefix    = config.GetPrefix();
+            preference = config.mPreference;
+            error      = kErrorNone;
+        }
+    }
+
+    return error;
```

Comment:
Should we fallback to `kWellknownPrefix` when not found in Network Data?

Reply:
What benefit does defaulting to `kWellknownPrefix` provide? A BR either needs to discover that a NAT64 prefix exists on the infrastructure or locally generate its own. I think we have to assume that the network data can hold at least one NAT64 prefix.

Reply:
The use of the hard-coded (well-known) prefix was just a temporary solution (mainly as a placeholder) till we agreed and finalized the mechanism for device to discover the NAT64 prefix (which we now have using Network Data external route entries).

If there are no NAT64 prefixes in network data, we return back `kErrorNotFound` (informing caller that address resolution failed and could not find any address).

Reply:
The obvious benefit of defaulting to `64:ff9b::/96` is that if the NAT64 prefix is indeed `64:ff9b::/96`, which is very likely, we can save the Network Data space by not adding it. 

But this will require wide support at the end-device side.

Reply:
I think we should require NAT64 prefix in the network data for clients to make use of NAT64 service. Lack of NAT64 prefix is an explicit statement that NAT64 service is not available.